course3/homework-string/my_string.cc

```diff
@@ -0,0 +1,192 @@
+#include "my_string.h"
+
+#include <cstring>
+#include <memory>
+
+MyString::MyString() 
+	:
+	mBuffer(nullptr)
+{
+}
+
+MyString::MyString(const MyString& that)
+{
+	if (that.empty())
+	{
+		mBuffer = nullptr;
+	}
+	else
+	{
+		mBuffer = std::make_unique<char[]>(std::strlen(that.mBuffer.get()) + 1);
+		std::strcpy(mBuffer.get(), that.mBuffer.get());
+	}
+}
+
+MyString::MyString(MyString&& that)
+{
+	mBuffer.swap(that.mBuffer);
+}
+
+MyString::MyString(const char* str)
+{
+	if (str == nullptr || str[0] == '\0')
+	{
+		mBuffer = nullptr;
+	}
+	else
+	{
+		mBuffer = std::make_unique<char[]>(std::strlen(str) + 1);
+		std::strcpy(mBuffer.get(), str);
+	}
+}
+
+MyString& MyString::operator=(const MyString& that)
+{
+	if (this == &that)
+		return *this;
+	
+	if (that.empty())
+	{
+		std::unique_ptr<char[]> oldBuffer = nullptr;
+		mBuffer.swap(oldBuffer);
```

Comment:
these two lines are a very elaborate way of saying
```
mBuffer.reset();
```
Here's some doco: http://en.cppreference.com/w/